peep.c

```diff
@@ -3839,6 +3839,65 @@ Perl_rpeep(pTHX_ OP *o)
                     }
                 }
             }
+            OP* rhs = cBINOPx(o)->op_first;
+            OP* lval = cBINOPx(o)->op_last;
+
+            /* Combine a simple SASSIGN OP with a PADSV lvalue child OP
+             * into a single OP. The RHS has op_next/op_sibparent pointers
+             * and op_moresib flags updated to match. */
+
+            /* This optimization covers arbitrarily complicated RHS OP
+             * trees. Separate optimizations may exist for specific,
+             * single RHS OPs, such as:
+             * "my $foo = undef;" or "my $bar = $other_padsv;" */
+
+            if (!(o->op_private & (OPpASSIGN_BACKWARDS|OPpASSIGN_CV_TO_GV))
+                 && lval && (lval->op_type == OP_PADSV)
+               ) {
+
+                /* SASSIGN's bitfield flags, such as op_moresib and
+                 * op_slabbed, will be carried over unchanged. */
+                OpTYPE_set(o, OP_PADSV_STORE);
+
+                /* Explicitly craft the new OPs op_flags, carrying
+                 * some bits over from the SASSIGN */
+                o->op_flags = (
+                    OPf_MOD | OPf_KIDS | OPf_STACKED |
+                    (o->op_flags & (OPf_WANT|OPf_PARENS))
+                );
+
+                /* Reset op_private flags */
+                o->op_private = 0;
+
+                /* Take relevant private flasgs from the PADSV */
+                o->op_private = (lval->op_private &
+                                (OPpLVAL_INTRO|OPpPAD_STATE|OPpDEREF));
+
+                /* Take the targ from the PADSV */
+                o->op_targ = lval->op_targ;
+
+                /* Fixup RHS op_sibparent pointer and op_moresib flag */
+                OpLASTSIB_set(rhs, o);
+
+                /* Fixup op_next ptrs */
+                /* oldoldop can be arbitrarily deep in the RHS OP tree */
+                oldoldop->op_next = o;
+                /* Even when (rhs != oldoldop), rhs might still have a
+                 * relevant op_next ptr to lval. This is definitely true
+                 * when rhs is OP_NULL with a LOGOP kid (e.g. orassign).
+                 * There may be other cases. */
+                if (rhs->op_next == lval)
+                    rhs->op_next = o;
```

Comment:
Ah; if this is the case (and yes the weird branch structure in LOGOPs make it complicated), then it's best not to free the op. Instead use

```c
op_null(lval);
```

To turn the op into one of those "ex-padsv" things you see in Concise output sometimes. Being a OP_NULL'ed op the optree finalizer will skip over it in all the actual op chaining.

Reply:
I did try `op_null(lval);` originally, but had problems with memory leakage. (Presumably because it was not linked to elsewhere in the optree?)  I'll give it another go; two questions:

1. Does this mean that `OP_PADSV_STORE `should remain a BINOP, not be a UNOP? 
2. If so, do I still need to do all of the fixing up of pointers that is in the current iteration of _peep.c_, or will e.g. some of the `op_next` pointers be automatically updated by the optree finalizer?

Reply:
> 2\. If so, do I still need to do all of the fixing up of pointers that is in the current iteration of _peep.c_, or will e.g. some of the `op_next` pointers be automatically updated by the optree finalizer?

That doesn't seem to happen. The OP_PADSV looks to still be in the `op_next` chain despite being a null op. e.g.
```
$ ./perl -Dt -e 'my $boop = 1'

EXECUTING...

(-e:0)	enter
(-e:0)	nextstate
(-e:1)	const(IV(1))
(-e:1)	null
(-e:1)	padsv_store
(-e:1)	leave e.g.
```
So seems like I still have to fix up the op_next pointers manually. In which case I'm not clear on why `op_null(lval)` is more desirable than `op_clear(lval)`, unless there are potentially some more pointers in the RHS that might point to the ex-PADSV and potentially not redirected by the optimization.

Reply:
I've pushed a commit that leaves OP_PADSV_STORE as a BINOP and does `op_null(lval)`, but still has the manual `op_next` fix-ups. See what you think?

Reply:
> * Does this mean that `OP_PADSV_STORE `should remain a BINOP, not be a UNOP?

Not quite sure I follow the question. `regen/opcodes` should call it an UNOP. If you manually create one you should `newUNOP(OP_PADSV_STORE, ...)`.

But if you took an existing BINOP and did `OpTYPE_set(o, OP_PADSV_STORE)` that is fine. A BINOP is larger in memory than an UNOP. All that means is that we're now storing an UNOP in a little memory slab that was accidentally too large for it. No huge problem.

> * If so, do I still need to do all of the fixing up of pointers that is in the current iteration of _peep.c_, or will e.g. some of the `op_next` pointers be automatically updated by the optree finalizer?

You can leave it in memory where it is, so this is all fine. But to answer the question: nothing in the finalizer will *change* any existing op_next pointers (apart from to unthread any `OP_NULL`s). If you did change something and it's already been threaded up you'll have to go chasing all the pointers yourself. Which is why it's usually best to change an op inplace, so no pointer hunting required.

Reply:
Hmm, I'm probably not following correctly, but what I've observed is:
1. 
* If I make OP_PADSV_STORE an UNOP, call `OpTYPE_set(o, OP_PADSV_STORE)`, fixup `rhs` to have no sibs, and then `op_null(lval)`, there are memory leaks detected by ASAN. To get rid of the leaks, I had to `op_free(lval)`. (If I left `rhs` untouched, `rpeep` would barf.)
versus:
* If I make OP_PADSV_STORE a BINOP, call `OpTYPE_set(o, OP_PADSV_STORE)`, and then `op_null(lhs)`, that seems fine.

2. The finalizer definitely does not unthread the ex-PADSV `OP_NULL` from the `op_next` pointer. I have to do that manually. (I can't see any code in `finalize_op` that would do this. Maybe `op_prune_chain_head` does this for OP_NULLs in some but not all locations?)

Reply:
Ok, you were definitely right that it should work with OP_PADSV_STORE being an UNOP, which this PR now uses again, along with `op_null(lval)`. I think what must have gone wrong on my first attempts was that I though I had to manually fix-up the RHS, but that wasn't actually necessary and it cast the `lval` adrift, leading to the leakage.

I think this PR is ready to re-review.